### ISSUE TITLE
ceres::sqrt(T(0.0)) is crashing for ceres::Jet

### ISSUE THREAD

**Issue by punitsoni** [ceres-solver/ceres-solver#447]
Currently the ceres::sqrt() function is implemented as following for Jet.
```
// sqrt(a + h) ~= sqrt(a) + h / (2 sqrt(a))
template <typename T, int N> inline
Jet<T, N> sqrt(const Jet<T, N>& f) {
  const T tmp = sqrt(f.a);
  const T two_a_inverse = T(1.0) / (T(2.0) * tmp);
  return Jet<T, N>(tmp, f.v * two_a_inverse);
}
```

When using this function in an AutoDiff cost functor, if the argument for sqrt is 0. It crashes. Although this is a divide by zero error for Jet with the derivative calculation. Divide by zero is not expected when calculating a sqrt of 0 for a scalar datatype like double. This requires the developer to explicitly check there is no sqrt(0) used in cost function calculation. Is this the recommended solution?

An example for this can be computation of vector norm in cost function.

```
// axis is of type Vector3<Scalar>
const Scalar norm_sq =
        axis.x() * axis.x() + axis.y() * axis.y() + axis.z() * axis.z();
const Scalar norm = ceres::sqrt(norm_sq); // this crashes if norm_sq is 0
```

**Comment by TomLKoller** [comment]
An easy way to fix the error would be adding a small value to your norm so it can not be 0 since you still should return a value, when your norm is 0.
`const Scalar norm = ceres::sqrt(norm_sq+1e-8); // fixes the crash`

Anyway an assert in sqrt(Jet) would help a lot for debugging purposes.

**Comment by NikolausDemmel** [comment]
Does this actually crash (with which error) or just return inf or nan in the derivatives (which would be kind of correct, since the derivative is not usable/defined in this case)? Looking at the code it seems in does just divide / multiply doubles. I would have expected inf for 1/0 or nan from 0/0.

That way, to my understanding, you can still use the resulting value in a comparison, but if you use it in your residual computation, it will (correctly) fail.

**Comment by JzHuai0108** [comment]
Yeah, infinite derivative is actually correct for  d f / d a at a == 0 and f = sqrt(a). To deal with this singularity, I have to stop using f in further arithmetic computations, which will propagate the infinite derivative and may ruin the derivative-dependent optimizer.
